pkg/applier/inventory/inventory_converter.go

```diff
@@ -0,0 +1,201 @@
+	if fromObj == nil {
+		return nil, fmt.Errorf("unstructured object is nil")
+	}
+	if toInv == nil {
+		return nil, fmt.Errorf("UnstructuredInventory object is nil")
+	}
+	objStatusMap := map[object.ObjMetadata]actuation.ObjectStatus{}
+	for _, s := range toInv.GetObjectStatuses() {
+		objStatusMap[inventory.ObjMetadataFromObjectReference(s.ObjectReference)] = s
+	}
+	klog.V(4).Infof("converting UnstructuredInventory to ResourceGroup object")
+	klog.V(4).Infof("creating list of %d resources", len(toInv.GetObjectRefs()))
+	var objs []interface{}
+	for _, objMeta := range toInv.GetObjectRefs() {
+		klog.V(4).Infof("converting to object reference: %s", objMeta)
+		objs = append(objs, map[string]interface{}{
+			"group":     objMeta.GroupKind.Group,
+			"kind":      objMeta.GroupKind.Kind,
+			"namespace": objMeta.Namespace,
+			"name":      objMeta.Name,
+		})
+	}
+	klog.V(4).Infof("Creating list of %d resource statuses", len(toInv.GetObjectRefs()))
+	var objStatus []interface{}
+	for _, objMeta := range toInv.GetObjectRefs() {
+		status, found := objStatusMap[objMeta]
+		if found {
+			klog.V(4).Infof("converting to object status: %s", objMeta)
+			objStatus = append(objStatus, map[string]interface{}{
+				"group":     objMeta.GroupKind.Group,
+				"kind":      objMeta.GroupKind.Kind,
+				"namespace": objMeta.Namespace,
+				"name":      objMeta.Name,
+				"status":    "Unknown",
+				"strategy":  status.Strategy.String(),
+				"actuation": status.Actuation.String(),
+				"reconcile": status.Reconcile.String(),
+			})
+		}
+	}
+
+	toObj := fromObj.DeepCopy()
+	// Set Config Sync specific metadata on the object
+	core.SetLabel(toObj, metadata.ManagedByKey, metadata.ManagedByValue)
+	core.SetLabel(toObj, metadata.SyncNamespaceLabel, ic.syncNamespace)
+	core.SetLabel(toObj, metadata.SyncNameLabel, ic.syncName)
+	core.SetLabel(toObj, metadata.SyncKindLabel, ic.syncKind)
+	core.SetAnnotation(toObj, metadata.ResourceManagementKey, metadata.ResourceManagementEnabled)
+	core.SetAnnotation(toObj, metadata.StatusModeKey, ic.statusMode)
+
+	if len(objs) == 0 { // Unset resource fields
+		klog.V(4).Infoln("clearing inventory resources")
+		unstructured.RemoveNestedField(toObj.UnstructuredContent(),
+			"spec", "resources")
+		unstructured.RemoveNestedField(toObj.UnstructuredContent(),
+			"status", "resourceStatuses")
+	} else { // Update resource fields
+		klog.V(4).Infof("storing inventory (%d) resources", len(objs))
+		err := unstructured.SetNestedSlice(toObj.UnstructuredContent(),
+			objs, "spec", "resources")
+		if err != nil {
+			return nil, err
+		}
+		err = unstructured.SetNestedSlice(toObj.UnstructuredContent(),
+			objStatus, "status", "resourceStatuses")
+		if err != nil {
+			return nil, err
+		}
+	}
+	generation := toObj.GetGeneration() // Update status.observedGeneration
+	klog.V(4).Infof("setting observedGeneration %d: ", generation)
+	err := unstructured.SetNestedField(toObj.UnstructuredContent(),
```

Comment:
Is updating status.observedGeneration here actually useful? I know the old code did, but IIUC, it will need to again be updated after the spec update and before the status update, right? Does the UnstructuredInventoryClient handle that?

Reply:
This function is called again after the spec update and before the status update, so this is valid. It could be considered redundant because the UnstructuredInventoryClient also updates the observedGeneration, but I considered this as just being defensive.